libcudacxx/include/cuda/std/__atomic/functions/cuda_ptx_derived.h

```diff
@@ -33,6 +31,250 @@ _LIBCUDACXX_BEGIN_NAMESPACE_STD
+
+template <class _Operand>
+using __cuda_atomic_enable_non_native_bitwise = __enable_if_t<_Operand::__size <= 16, bool>;
+
+template <class _Operand>
+using __cuda_atomic_enable_native_bitwise = __enable_if_t<_Operand::__size >= 32, bool>;
+
+template <class _Operand>
+using __cuda_atomic_enable_non_native_ld_st = __enable_if_t<_Operand::__size <= 8, bool>;
+
+template <class _Operand>
+using __cuda_atomic_enable_native_ld_st = __enable_if_t<_Operand::__size >= 16, bool>;
+
+template <class _Type, class _Order, class _Operand, class _Sco, __cuda_atomic_enable_non_native_ld_st<_Operand> = 0>
+static inline _CCCL_DEVICE void
+__cuda_atomic_load(const _Type* __ptr, _Type& __dst, _Order, _Operand, _Sco, __atomic_cuda_mmio_disable)
+{
+  constexpr uint64_t __alignmask = (sizeof(uint16_t) - 1);
+  uint16_t* __aligned            = (uint16_t*) ((intptr_t) __ptr & (~__alignmask));
+  const uint8_t __offset         = uint16_t((intptr_t) __ptr & __alignmask) * 8;
+
+  uint16_t __value = 0;
+  __cuda_atomic_load(__aligned, __value, _Order{}, __atomic_cuda_operand_b16{}, _Sco{}, __atomic_cuda_mmio_disable{});
+
+  __dst = static_cast<_Type>(__value >> __offset);
+}
+
+template <class _Type, class _Order, class _Operand, class _Sco, __cuda_atomic_enable_non_native_bitwise<_Operand> = 0>
+static inline _CCCL_DEVICE bool
+__cuda_atomic_compare_exchange(_Type* __ptr, _Type& __dst, _Type __cmp, _Type __op, _Order, _Operand, _Sco)
+{
+  constexpr uint64_t __alignmask = (sizeof(uint32_t) - 1);
+  constexpr uint32_t __sizemask  = (1 << (sizeof(_Type) * 8)) - 1;
+  uint32_t* __aligned            = (uint32_t*) ((intptr_t) __ptr & (~__alignmask));
+  const uint8_t __offset         = uint32_t((intptr_t) __ptr & __alignmask) * 8;
+  const uint32_t __valueMask     = __sizemask << __offset;
+  const uint32_t __windowMask    = ~__valueMask;
+  const uint32_t __cmpOffset     = __cmp << __offset;
+  const uint32_t __opOffset      = __op << __offset;
+
+  // Algorithm for 8b CAS with 32b intrinsics
+  // __old = __window[0:32] where [__cmp] resides within some offset.
+  uint32_t __old;
+  // Start by loading __old with the current value, this optimizes for early return when __cmp is wrong
+  NV_IF_TARGET(
+    NV_PROVIDES_SM_70,
+    (__cuda_atomic_load(
+       __aligned, __old, __atomic_cuda_relaxed{}, __atomic_cuda_operand_b32{}, _Sco{}, __atomic_cuda_mmio_disable{});),
+    (__cuda_atomic_load(
+       __aligned, __old, __atomic_cuda_volatile{}, __atomic_cuda_operand_b32{}, _Sco{}, __atomic_cuda_mmio_disable{});))
+  // Reemit CAS instructions until we succeed or the old value is a mismatch
+  while (__cmpOffset == (__old & __valueMask))
+  {
+    // Combine the desired value and most recently fetched expected masked portion of the window
+    const uint32_t __attempt = (__old & __windowMask) | __opOffset;
+
+    if (__cuda_atomic_compare_exchange(
+          __aligned, __old, __old, __attempt, _Order{}, __atomic_cuda_operand_b32{}, _Sco{}))
```

Comment:
This will generate release/acquire/acq_rel/seq_cst fences within the inner loop, but it'd suffice to issue the right fences before/after the loop.

Reply:
Are you mistaking the `__atomic_cuda_compare_exchange` frontend with this call? This dispatches purely to PTX.